delivery_cttexpress/models/cttexpress_request.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2022 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from lxml import etree
+
+import logging
+
+from zeep import Client
+from zeep.helpers import serialize_object
+from zeep.plugins import HistoryPlugin
+
+
+_logger = logging.getLogger(__name__)
+
+CTTEXPRESS_API_URL = {
+    "test": "http://iberws.tourlineexpress.com:8686/ClientsAPI.svc?singleWsdl",
+    "prod": "http://iberws.tourlineexpress.com:8700/ClientsAPI.svc?singleWsdl",
+}
+
+
+def log_request(method):
+    """Decorator to write raw request/response in the CTT request object"""
+    def wrapper(*args, **kwargs):
+        res = method(*args, **kwargs)
+        try:
+            args[0].ctt_last_request = etree.tostring(
+                args[0].history.last_sent["envelope"],
+                encoding="UTF-8",
+                pretty_print=True,
+            )
+            args[0].ctt_last_response = etree.tostring(
+                args[0].history.last_received["envelope"],
+                encoding="UTF-8",
+                pretty_print=True,
+            )
+        # Don't fail hard on this. Sometimes zeep can't keep history
+        except Exception:
+            return res
+        return res
+    return wrapper
+
+
+class CTTExpressRequest:
+    """Interface between CTT Express SOAP API and Odoo recordset.
+       Abstract CTT Express API Operations to connect them with Odoo
+    """
+    def __init__(self, user, password, agency, customer, contract, prod=False):
+        self.user = user
+        self.password = password
+        self.agency = agency
+        self.customer = customer
+        self.contract = contract
+        self.history = HistoryPlugin(maxlen=10)
```

Comment:
I prefer to simplify it (e.g. https://github.com/OCA/delivery-carrier/blob/13.0/delivery_tnt_oca/models/tnt_request.py#L20-L22).

Reply:
IMO the connector class should be completely abstract. In its current form it could even be packaged as a generic connector library for CTT Express ready to use in any python project (and maybe we should do that to ease maintainance :thinking: )